buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -545,11 +562,65 @@ static int writeUtf8(AbstractByteBuf buffer, int writerIndex, CharSequence seq,
         return writerIndex - oldWriterIndex;
     }
 
+    /**
+     * Returns max bytes length of UTF8 character sequence of the given length.
+     */
+    public static int utf8MaxBytes(final int seqLength) {
+        return seqLength * MAX_BYTES_PER_CHAR_UTF8;
+    }
+
     /**
      * Returns max bytes length of UTF8 character sequence.
+     * <p>
+     * It behaves like {@link #utf8MaxBytes(int)} applied to {@code seq} {@link CharSequence#length()}.
      */
     public static int utf8MaxBytes(CharSequence seq) {
-        return seq.length() * MAX_BYTES_PER_CHAR_UTF8;
+        return utf8MaxBytes(seq.length());
+    }
+
+    /**
+     * Returns the exact bytes length of UTF8 character sequence.
+     * <p>
+     * This method is producing the exact length only with well-formed pairs, otherwise
+     * it will not perform underrated estimations according to {@link #writeUtf8(ByteBuf, CharSequence)}.
+     */
+    public static int utf8Bytes(final CharSequence seq) {
+        int seqLength = seq.length();
+        // the algorithm is just adding the additional length of not ASCII chars to the baseline ASCII length
+        int encodedLength = seqLength;
+        int i = 0;
+        // ASCII
+        while (i < seqLength && seq.charAt(i) < 0x80) {
+            ++i;
+        }
+        if (i < seqLength) {
+            // !ASCII
+            encodedLength += additionalUtf8Bytes(seq, i, seqLength);
+            // integer overflow on additionalUtf8Bytes
+            if (encodedLength < seqLength) {
+                throw new IllegalArgumentException("UTF-8 encoded length overflow");
+            }
+        }
+        return encodedLength;
+    }
+
+    private static int additionalUtf8Bytes(final CharSequence sequence, final int start, final int length) {
+        int encodedLength = 0;
+        for (int i = start; i < length; i++) {
+            final int c = sequence.charAt(i);
+            // making it 100% branchless isn't rewarding due to the many bit operations necessary!
+            if (c < 0x800) {
+                // branchless version of: c <= 127 ? 0:1
+                encodedLength += (0x7f - c) >>> 31;
+            } else {
+                encodedLength += 2;
+                if (MIN_SURROGATE <= c && c <= MAX_SURROGATE) {
+                    assert Character.codePointAt(sequence, i) != c : "Not well formed pair";
```

Comment:
our encoding method `writeUtf8` accounts for invalid utf8 encodings just like the JDK. This may result in writing an `WRITE_UTF_UNKNOWN` if the encoding is not correct, but still proceed without an exception. So this method may not support some methods which are supported by `writeUtf8`. We should consider the following:
- change from `assert` to a runtime check. otherwise the encoding size may not be exact, and not detected if asserts for this package are not enabled.
- be more lenient about encoding issues and calculate the size according to `writeUtf8` (and the default JDK configuration).
- add a mode which can conditionally switch between the previous 2 approaches